addon/-private/system/relationships/state/has-many.js

```diff
@@ -111,6 +111,32 @@ export default class ManyRelationship extends Relationship {
     super.removeCanonicalInternalModelFromOwn(internalModel, idx);
   }
 
+  unloadInverseInternalModelFromOwn(internalModel) {
+    super.unloadInverseInternalModelFromOwn(internalModel);
+
+    const canonicalIndex = this.canonicalState.indexOf(internalModel);
+
+    if (canonicalIndex !== -1) {
+      this.canonicalState.splice(canonicalIndex, 1);
+    }
+
+    const manyArray = this._manyArray;
+
+    if (manyArray) {
+      const currentState = manyArray.currentState;
+      const idx = currentState.indexOf(internalModel);
+
+      if (idx !== -1) {
+        currentState.splice(idx, 1);
+
+        manyArray.set('length', currentState.length);
+        manyArray.notifyPropertyChange('length');
+
+        this.internalModel.notifyHasManyRemoved(this.key, internalModel, idx);
```

Comment:
> the Enumerable docs suggest just ...

Enumerable docs appear in-complete `arrayContent{Will,Did}Change` stuff seems to be missing.

> I'm still a bit confused why this (1) didn't seem to be an issue unless there was the extra proxy from the async wrapper

As soon as there is a ArrayProxy (in this case: PromiseManyArray) `length` had the chance of getting out of sync.